File: leetcode/leetcode_py/convert-sorted-array-to-binary-search-tree.py
'''
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:        
        def buildNode(l, r):
            if r < l or l > r:
                return None
            m = int((l+r)/2)
            node = TreeNode(nums[m])
            node.left = buildNode(l, m-1)
            node.right = buildNode(m+1, r)
            return node
        return buildNode(0, len(nums) - 1)
        

File: leetcode/leetcode_cpp/convert-sorted-array-to-binary-search-tree.cpp
/*
https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree

Solution1. BST recursion
find middle and put it on root of others (preorder)

def bst(l, r, nums):
    if (..) return none
    m = (l+r)/2
    node = node(nums[m])
    node.left = bst(l, m-1)
    node.right = bst(m + 1, r)
    return node
bst ...

*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* bst(vector<int>& nums, int l, int r) {
        if (r < l || l > r) return nullptr;
        int m = (l + r) / 2;
        TreeNode* node = new TreeNode(nums[m]);
        node->left = bst(nums, l, m - 1);
        node->right = bst(nums, m + 1, r);
        return node;
    }

    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return bst(nums, 0, nums.size() - 1);
    }
};